### ISSUE TITLE
Building on macOS does not appear to work

### ISSUE THREAD

**Issue by djmarcin** [grailbio/bazel-toolchain#83]
I made a very basic cc_binary example in https://github.com/djmarcin/llvm-toolchain-test, but the gist is that none of the headers can be found while building on macOS 11.4, but builds perfectly on Linux.

It appears that the macOS sysroot is being properly detected (sysroot=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk), and system headers do exist there, but it still isn't able to compile correctly.

The generated command for me is:
```
(cd /private/var/tmp/_bazel_djmarcin/4253e2a12d8dd57e508832c54036fb7e/execroot/example && \
  exec env - \
    PATH=/bin:/usr/bin:/usr/local/bin \
    PWD=/proc/self/cwd \
  external/llvm_toolchain/bin/cc_wrapper.sh -U_FORTIFY_SOURCE -fstack-protector -fno-omit-frame-pointer -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign '-std=c++17' '-stdlib=libc++' -MD -MF bazel-out/darwin-fastbuild/bin/_objs/hello/hello.pic.d -fPIC -iquote . -iquote bazel-out/darwin-fastbuild/bin -iquote external/bazel_tools -iquote bazel-out/darwin-fastbuild/bin/external/bazel_tools -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' '-fdebug-prefix-map=external/llvm_toolchain/=external/llvm_toolchain/' '--sysroot=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk' -c hello.cc -o bazel-out/darwin-fastbuild/bin/_objs/hello/hello.pic.o)
```

and it quickly errors out with a lot of messages like this (but many more)
```
In file included from hello.cc:1:
In file included from external/llvm_toolchain/bin/../include/c++/v1/iostream:37:
In file included from external/llvm_toolchain/bin/../include/c++/v1/ios:214:
In file included from external/llvm_toolchain/bin/../include/c++/v1/iosfwd:95:
external/llvm_toolchain/bin/../include/c++/v1/wchar.h:137:77: error: use of undeclared identifier 'wcschr'
wchar_t* __libcpp_wcschr(const wchar_t* __s, wchar_t __c) {return (wchar_t*)wcschr(__s, __c);}
                                                                            ^
external/llvm_toolchain/bin/../include/c++/v1/wchar.h:144:87: error: use of undeclared identifier 'wcspbrk'
wchar_t* __libcpp_wcspbrk(const wchar_t* __s1, const wchar_t* __s2) {return (wchar_t*)wcspbrk(__s1, __s2);}
                                                                                      ^
external/llvm_toolchain/bin/../include/c++/v1/wchar.h:151:78: error: use of undeclared identifier 'wcsrchr'; did you mean 'wcschr'?
wchar_t* __libcpp_wcsrchr(const wchar_t* __s, wchar_t __c) {return (wchar_t*)wcsrchr(__s, __c);}
                                                                             ^
external/llvm_toolchain/bin/../include/c++/v1/wchar.h:139:16: note: 'wcschr' declared here
const wchar_t* wcschr(const wchar_t* __s, wchar_t __c) {return __libcpp_wcschr(__s, __c);}
```

Running the command above directly with `-v` shows the resolved include search paths:
```
clang -cc1 version 11.0.0 based upon LLVM 11.0.0 default target x86_64-apple-darwin20.5.0
ignoring nonexistent directory "bazel-out/darwin-fastbuild/bin/external/bazel_tools"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/local/include"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/Library/Frameworks"
#include "..." search starts here:
 .
 bazel-out/darwin-fastbuild/bin
 external/bazel_tools
#include <...> search starts here:
 external/llvm_toolchain/bin/../include/c++/v1
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include/c++/v1
 external/llvm_toolchain/lib/clang/11.0.0/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/System/Library/Frameworks (framework directory)
End of search list.
```

This is pretty confusing because if I trace through these directories, wchar.h exists in ` external/llvm_toolchain/bin/../include/c++/v1
`, `{sysroot}/usr/include/c++/v1`, and `{sysroot}/usr/include`, the first two of which have `#include_next <wchar.h>` and the last containing the symbols that are supposedly missing. The first two files are in fact the same exact file, so perhaps only one of the two paths should exist in the search path?

**Comment by djmarcin** [comment]
I believe this is due to [this bug](https://reviews.llvm.org/D89001) which exists in 11.0.0. Updating to LLVM 12.0.0 fixes the issue. I'm not sure what the workaround would be if you needed to use LLVM 11.0.0, but different flags would likely need to be passed on macOS.

**Comment by siddharthab** [comment]
I think this is because the llvm compilation for 11.0 did not include the sysroot, and macOS Big Sur has these header files only in the Xcode sysroot paths. The fix for this is simple (see [this commit](https://github.com/grailbio/rules_r/commit/8b896504e5271bd0d1e87591e07e9f082582bf0e) in my other project).

Note that Homebrew LLVM does not have this problem because they recompiled LLVM with a default sysroot setting.

I will fix it here as soon as I get to this project. I have not been able to give this project the attention that it's due and things have been languishing. I apologize for this.

**Comment by djmarcin** [comment]
This looks like a different problem to me. This is happening at the compilation step not linking, so the extra `LDFLAGS` don't seem like they would make a difference. The reason Homebrew LLVM does not have this problem appears to be because they cherry-pick the fix to the linked issue before building https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/llvm@11.rb#L84-L87

One potential fix is to not set the sysroot flag at all and instead pass `-isystem${sysroot}/usr/include` in its place. This prevents the double inclusion of the `include/c++/v1` files. I don't know if there's a flag you can pass clang to prevent it from using its included copy of these files, but it doesn't look like it from my reading of the patch.

**Comment by siddharthab** [comment]
Thanks for the clarification. I thought it was just because of an unspecified sysroot, but looks like the problem is the opposite. The fix is not present in 11.1.0, so I suppose the only solution is to use LLVM 12.0.0+.

I suppose one way to resolve this issue would be to "hide" the header files in the toolchain for the version pairs where this problem could happen.

**Comment by siddharthab** [comment]
I think solving this could be out of scope for this project. I am going to close this issue.

Please feel free to reopen if you have an idea that can be implemented easily, but given that LLVM 12 already works, I would not want neither you nor I to spend too much time on it.
